R/fread.R

```diff
@@ -164,29 +169,61 @@ autostart=NA)
   if (check.names) {
     setattr(ans, 'names', make.names(names(ans), unique=TRUE))
   }
-  cols = NULL
-  if (stringsAsFactors)
-    cols = which(vapply(ans, is.character, TRUE))
-  else if (length(colClasses)) {
-    if (is.list(colClasses) && "factor" %chin% names(colClasses))
-      cols = colClasses[["factor"]]
-    else if (is.character(colClasses) && "factor" %chin% colClasses)
-      cols = which(colClasses=="factor")
-  }
-  setfactor(ans, cols, verbose)
-  # 2007: is.missing is not correct since default value of select is NULL
+
+  colClassesAs = attr(ans, "colClassesAs")   # should only be present if one or more are != ""
+  for (j in which(colClassesAs!="")) {       # # 1634
+    v <- .subset2(ans, j)
+    new_class = colClassesAs[j]
+    new_v <- tryCatch({    # different to read.csv; i.e. won't error if a column won't coerce (fallback with warning instead)
+      switch(new_class,
+             "factor" = as_factor(v),
+             "complex" = as.complex(v),
+             "raw" = as_raw(v),  # Internal implementation
+             "Date" = as.Date(v),
+             "POSIXct" = as.POSIXct(v),
```

Comment:
Is there any way to build customized timezones into the API? Otherwise we should use `tz = 'UTC'` by default? Since the same string on different machines will parse differently otherwise...

Reply:
Good idea:  could add `tz=` to fread I guess.  But shouldn't default be whatever `as.POSIXct()` does by default (which is local time iirc) ?

Reply:
I'm basically arguing against R default behavior since I've always found it a bit murky to use inferred timezones --> `fread` behaves differently on different machines.

Adding `tz` to `fread` feels a bit strange, maybe accept `tz` as a "class" within `colClasses` a la

`colClasses = list(POSIXct = 'time', tz = 'UTC')`

Or (requires a bit more parsing within our code but I think feels pretty natural):

`colClasses = list(POSIXct[UTC] = 'time')`

Reply:
But eventually `colClasses` won't be needed to get a `POSIXct` (it would be built in as native type). Hence `tz=` being argument to `fread`.

Reply:
And I guess `tz` argument could have an API like `colClasses` to allow flexibly specifying multiple time zones:

`tz = list(UTC = c('start', 'end'), 'Australia/South' = c('start_local', 'end_local'))`

and

`tz = c(start = 'UTC', end = 'UTC', start_local = 'Australia/South', end_local = 'Australia/South')`

Reply:
I think it make sense to use `tz="UTC"` and just document that

Reply:
my only hesitation is I'm not _100%_ sure we can do this before implementing the tz argument... e.g. some times only exist in certain time zones, so conversion may fail if we assume UTC but could read on user's machine otherwise?

Reply:
if it may fail then `tryCatch`, providing more complex structure to `colClasses` is good idea but should be added with care (ideally another PR), ideally aligning to API we will use in csvy specification.